core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,404 @@
+	} else {
+		dpair = clk_div_glitch_disabled;
+		dpair_sz = ARRAY_SIZE(clk_div_glitch_disabled);
+	}
+
+	if (clock_divisor > dpair[dpair_sz - 1].divisor) {
+		ibfd = dpair[dpair_sz - 1].ibfd;
+	} else {
+		for (n = 0; n < dpair_sz; n++) {
+			if (dpair[n].divisor >= clock_divisor) {
+				ibfd = dpair[n].ibfd;
+				break;
+			}
+		}
+	}
+
+	return ibfd;
+}
+
+TEE_Result i2c_init(struct ls_i2c_data *i2c_data)
+{
+	struct i2c_regs *regs = NULL;
+	uint16_t clock_divisor = 0;
+	uint8_t ibfd = 0; /* I2c Bus Frequency Divider Register */
+	size_t size = 0;
+	int node = 0;
+	vaddr_t ctrl_base = 0;
+	void *fdt = NULL;
+
+	/*
+	 * First get the I2C Controller base address from the DTB
+	 * if DTB present and if the I2C Controller defined in it.
+	 */
+	fdt = get_embedded_dt();
+	if (!fdt) {
+		EMSG("Unable to get the Embedded DTB, I2C init failed\n");
+		return TEE_ERROR_GENERIC;
+	}
+
+	node = fdt_path_offset(fdt,
+			       i2c_controller_map[i2c_data->i2c_controller]);
+	if (node > 0) {
+		if (dt_map_dev(fdt, node, &ctrl_base, &size) < 0) {
+			EMSG("Unable to get virtual address");
+			return TEE_ERROR_GENERIC;
+		}
+	} else {
+		EMSG("Unable to get I2C offset node");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	i2c_data->base = ctrl_base;
+
+	regs = (struct i2c_regs *)ctrl_base;
+
+	clock_divisor = (i2c_data->i2c_bus_clock + i2c_data->speed - 1) /
+			i2c_data->speed;
+	ibfd = i2c_get_ibfd(ctrl_base, clock_divisor);
+
+	io_write8((vaddr_t)&regs->ibfd, ibfd);
+
+	i2c_reset(ctrl_base);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_bus_test_bus_busy(struct i2c_regs *regs, bool test_busy)
+{
+	unsigned int n = 0;
+	uint8_t reg = 0;
+
+	for (n = 0; n < I2C_NUM_RETRIES; n++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBAL) {
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			return TEE_ERROR_BUSY;
+		}
+
+		if (test_busy && (reg & I2C_IBSR_IBB))
+			break;
+
+		if (!test_busy && !(reg & I2C_IBSR_IBB))
+			break;
+
+		mdelay(1);
+	}
+
+	if (n == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_transfer_complete(struct i2c_regs *regs, bool test_rx_ack)
+{
+	unsigned int n = 0;
+	uint8_t reg = 0;
+
+	for (n = 0; n < I2C_NUM_RETRIES; n++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBIF) {
+			/* Write 1 to clear the IBIF field */
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			break;
+		}
+		mdelay(1);
+	}
+
+	if (n == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	if (test_rx_ack && (reg & I2C_IBSR_RXAK))
+		return TEE_ERROR_NO_DATA;
+
+	if (reg & I2C_IBSR_TCF)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_read(struct i2c_regs *regs, unsigned int slave_address,
+			   struct i2c_operation *operation,
+			   bool is_last_operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int n = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr, (slave_address << BIT(0)) | BIT(0));
```

Comment:
`slave_address << BIT(0)` looks a bit funny, why are you using `BIT(0)` here?

Reply:
No specific reason to use this, will update.